<a href="https://colab.research.google.com/github/gyhfkr/PRAKTIKUN-SOCIAL-NETWORK-ANALYSIS/blob/main/21110024_Gayuh_Fikri_Khusnaeli_Preprocessing_SNA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GAYUH FIKRI KHUSNAELI
# 21110024
# S1-SD-02-A

In [1]:
#import modul
import pandas as pd
import numpy as np
import re
import re as reg
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
# Membaca data dari file Excel
data=pd.read_excel("ReviewTerbaru.xlsx", sheet_name='Lokawisata Baturraden')
data

Nama          Waktu     Rating  \
0              Sonia Soraya  3 minggu lalu  5 bintang   
1          Andy Kristyagita   sebulan lalu  5 bintang   
2               Edza Tamara  2 minggu lalu  5 bintang   
3               Abrakadabra  seminggu lalu  5 bintang   
4              Riko Tampati   sebulan lalu  4 bintang   
..                      ...            ...        ...   
285             bunda popon   3 bulan lalu  5 bintang   
286                  petruk   setahun lalu  5 bintang   
287             Eva Meliana   2 tahun lalu  5 bintang   
288  Deandra Almerrafif W.T   sebulan lalu  5 bintang   
289            agniya ridha   2 tahun lalu  5 bintang   

                                                Review  
0    Bismillah tadi kita ke ke sana..  Pas masuk di...  
1    Walaupun sudah berulang kali ke sini semenjak ...  
2    Suasanan malam hari di lokawisata baturraden n...  
3    Tempat nya sangat Bersih Dan rapih, butuh stam...  
4    Masih masa lebaran, tiket masuk agak mahal 25r...  
..                                                 ...  
285  Paket komplit, murah meriah cocok untuk acara ...  
286  Lokasi nya nyaman dan aman, banyak tempat untu...  
287  Tempat nya bagus, sejuk, pemandian nya air nya...  
288           View bagus,,, tiket murah, berenang free  
289  Seneng banget bisa kesini, suasananya alam ban...  

[290 rows x 4 columns]

# PREPROCESSING

In [21]:
!pip install Sastrawi
from Sastrawi.StopWordRemover.StopWordRemoverFactory import StopWordRemoverFactory
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

# Preprocessing
slangs = {'yg': 'yang', 'tdk': 'tidak', 'pd': 'pada', 'mlh': 'malah',
          'jgn': 'jangan', 'jg': 'juga', 'tp': 'tapi', 'blkg': 'belakang',
          'dr': 'dari', 'klo': 'kalo', 'lg': 'lagi'}
preprocessing_review = []

for sentence in data['Review']:
    # Hapus semua karakter khusus
    processed_comment = re.sub(r'\W', ' ', str(sentence))

    # Konversi menjadi huruf kecil
    processed_comment = processed_comment.lower()

    # Hapus angka
    processed_comment = re.sub(r'\d+', ' ', processed_comment)

    # Hapus kata-kata yang terdiri dari satu, dua, atau tiga huruf
    processed_comment = re.sub(r'\b\w{1,3}\b', '', processed_comment)

    # Hapus karakter duplikat
    pattern = re.compile(r"(.)\1{1,}", re.DOTALL)
    processed_comment = pattern.sub(r"\1", processed_comment)

    # Koreksi kata-kata slang
    words = processed_comment.split()
    rfrm = [slangs[word] if word in slangs else word for word in words]
    processed_comment = " ".join(rfrm)

    # Hapus stopword
    factory = StopWordRemoverFactory()
    more_stopword = ['tak', 'jd', 'per', 'nya']  # Menambahkan stopword
    stopwords = factory.get_stop_words() + more_stopword
    temp = [t for t in re.findall(r'\b[a-z]+-?[a-z]+\b', processed_comment) if t not in stopwords]
    processed_comment = ' '.join(temp)

    # Stemming
    stemmer = StemmerFactory().create_stemmer()
    processed_comment = stemmer.stem(processed_comment)

    # Ganti beberapa spasi dengan satu spasi
    processed_comment = re.sub(r'\s+', ' ', processed_comment, flags=re.I)

    # Tambahkan komentar yang memiliki lebih dari 3 huruf ke dalam list
    if len(processed_comment.split()) > 3:
        preprocessing_review.append(processed_comment)

# Buat DataFrame untuk data yang telah diproses
preprocessing_review = pd.DataFrame({'Processed_Review': preprocessing_review})

In [22]:
# Tampilkan DataFrame hasil preprocessing
print(preprocessing_review)

                                      Processed_Review
0    bismilah tadi sana masuk suru cuci tangan paka...
1    walaupun ulang kali sini semenjak kecil rasa b...
2    suasanan malam hari lokawisata baturaden nih s...
3    tempat sangat bersih rapih butuh stamina utama...
4    masa lebaran tiket masuk mahal sangat ramai ba...
..                                                 ...
262      paket komplit murah riah cocok acara keluarga
263  lokasi nyaman aman banyak tempat berswaphoto b...
264  tempat bagus sejuk mandi dingin hangat flaying...
265                  view bagus tiket murah renang fre
266  neng banget kesini suasana alam banget bukit s...

[267 rows x 1 columns]


# MENYIMPAN KE EXCEL

In [26]:
# Simpan data yang telah diproses ke dalam file Excel
preprocessing_review.to_excel("Processed_Review.xlsx", index=False)

In [27]:
# Jumlah record/dokumen sebelum dan sesudah pembersihan data
jumlah_record_sebelum = len(data)
jumlah_record_sesudah = len(preprocessing_review)
print("Jumlah record/dokumen sebelum pembersihan data:", jumlah_record_sebelum)
print("Jumlah record/dokumen setelah pembersihan data:", jumlah_record_sesudah)

Jumlah record/dokumen sebelum pembersihan data: 290
Jumlah record/dokumen setelah pembersihan data: 267


# INTERPRETASI
Dari hasil preprocessing tersebut dapat kita ketahui setelah melakukan analisis, terungkap bahwa sebelum dilakukan pembersihan data, terdapat 290 record atau dokumen dalam dataset. Namun, setelah proses pembersihan dilakukan, jumlah record tersebut berkurang menjadi 267. Penurunan jumlah ini menandakan bahwa beberapa record atau dokumen telah dihapus selama proses pembersihan data. Langkah pembersihan data tersebut bertujuan untuk meningkatkan kualitas dataset dengan menghilangkan record atau dokumen yang tidak relevan atau tidak informatif, seperti yang mungkin terjadi pada record-record yang hanya terdiri dari satu, dua, atau tiga huruf saja. Dengan demikian, dapat disimpulkan bahwa proses pembersihan data telah berhasil meningkatkan kebersihan dan kegunaan dataset dengan mengurangi data yang tidak berguna.